Fix: `ActiveRecord::Calculations#ids` returns duplicate ids

joshuay03 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created to address https://github.com/rails/rails/issues/46455.

### Detail

This Pull Request changes `ActiveRecord::Calculations#ids` to only return the unique ids of the base model, ignoring the join on eager loaded, preloaded and included associations.

The cause for this was the join dependancy applied within `#pluck` when the relation contained preloads. `#ids` has been updated to query the relation's foreign key itself rather than relying on `#pluck` unless the records have already been loaded.

### Additional information

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [x] CI is passing.

joshuay03 committed:
[Fix: #46455] ActiveRecord::Calculations#ids plucks included associations IDs

adrianna-chang-shopify reviewed (commented):
👋 Hey Joshua, thanks for picking this up!

Relying on the base relation makes sense to me, but I think `#ids` should still rely on `#pluck`, instead of extracting the pluck logic to `#perform_pluck` and calling that directly from `#ids`. `#pluck` avoids making queries if the relation is already loaded:
```ruby
def pluck(*column_names)
  if loaded? && all_attributes?(column_names)
    result = records.pluck(*column_names)
    if @async
      return Promise::Complete.new(result)
    else
      return result
    end
  end
...
end
```

So if we jump straight to the `#perform_pluck` logic you've extracted when `#ids` is called, we will always perform a query. What do you think about still using `#pluck` in `#ids`, but scoping to the base relation there, ie.
```ruby
def ids
  klass.pluck(primary_key)
end
```
?

adrianna-chang-shopify commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,19 @@
+*   Exclude ids of included records with #ids
```
```suggestion
*   Return only unique ids from ActiveRecord::Calculations#ids
```
We should clarify that we're excluding duplicated ids here (in other words, returning only unique ids)

joshuay03 replied:
I think my tests may have been a bit misleading. To reiterate:

> This Pull Request changes ActiveRecord::Calculations#ids to only pluck the ids of the base model, ignoring eager loaded, preloaded and included associations.

I'm recommending that we ignore preloaded record ids altogether. So it's not just excluding duplicates, it's **only** plucking the base model's ids. This may not be apparent from the tests since the associations created using `5.times` would probably contain ids matching existing base model records.

adrianna-chang-shopify replied:
But the point of not applying the join dependency is so that we don't get back the same id (for the base relation) more than once, right? Before these changes (using your tests as an example), we'd end up with a relation containing 5 Active Records for Company with id 1, with each record tied to a Contract. We're not excluding the ids on the `Contract` records, we're excluding the duplicated ids of the base relation (Company).

joshuay03 replied:
Ah yes of course I see what you mean now, that is correct. I think having been away from this for a while I managed to forget what the actual issue was haha. I have actioned your suggestion!

joshuay03 commented:
Thanks for the review! @adrianna-chang-shopify

> So if we jump straight to the #perform_pluck logic you've extracted when #ids is called, we will always perform a query. What do you think about still using #pluck in #ids, but scoping to the base relation there, ie.

This makes way more sense. I'm still getting used to how AR base objects are structured so thanks for the suggestion which I've now actioned.

adrianna-chang-shopify commented:
No problem! Test failure looks legit though, let me take a look 🤔

joshuay03 commented:
@adrianna-chang-shopify I've updated to pluck on `records` instead since plucking on `klass` introduced a regression where scoping was ignored. I have added a failing test which passed before this PR: https://github.com/rails/rails/compare/1114ff53f5a64b520d2e76b8fa1fc069cf4c8a85..a1333ca0bddde73380418bd99819289298f65d86.

casperisfine commented on activerecord/lib/active_record/relation/calculations.rb:
```diff
@@ -317,12 +317,12 @@ def async_pick(*column_names)
     #   Person.joins(:companies).ids # SELECT people.id FROM people INNER JOIN companies ON companies.person_id = people.id
     def ids
-      pluck primary_key
+      records.pluck(primary_key)
```
`records` call `load`, so this impact performance in two ways:

  - It will perform a `SELECT *`. so fetch more data.
  - It will instantiate a model for each row.

So I don't think this solution is acceptable.

joshuay03 replied:
Thanks for the review @casperisfine.
I'll see if I can find a better solution.

adrianna-chang-shopify commented:
Ah, yes scoping would be a problem 🤦‍♀️ Is there a reason not to do `pluck(primary_key).uniq`? That seems simplest..

casperisfine commented:
> Is there a reason not to do `pluck(primary_key).uniq`? That seems simplest..

That is doing the deduplication on the Ruby side, that isn't ideal. Something like `SELECT DISTINCT <pk>` might be a solution, but not 100% sure I haven't dug much in the original issue yet.

adrianna-chang-shopify commented:
Ah, yeah that makes sense to me 👍 `distinct.pluck(primary_key)` gives us what we want, right? Although we're still constructing the join dependency unnecessarily at that point, so perhaps it makes more sense to inline the query inside `#ids` instead of relying on `#pluck`..?

joshuay03 commented:
> so perhaps it makes more sense to inline the query inside `#ids` instead of relying on `#pluck`..?

I agree @adrianna-chang-shopify. I've updated `#ids` to pluck if records have been preloaded, and select the foreign key otherwise, also handing where contradictions and async just like `#pluck` and `#pick`. Changes and additional test coverage visible here: https://github.com/rails/rails/compare/f962068a6eb37581682f94edf76b1dbaad773ea8..307816734ee0122380fdc8a74156c82872501193.

byroot commented:
Nice work! Thank you.

simi commented:
:thinking: this adds `async_ids` as well, which is missing in commit message and changelog. Is that on purpose (to not promote async methods yet)?

matthewd commented:
I'm not sure this is something we should / need to change. But if we're going to, I think we need to take some more inspiration from the existing pluck tests, as the new implementation is no longer covered by them:

```ruby
  def test_ids_if_table_included
    c = Company.create!(name: "test", contracts: [Contract.new(developer_id: 7)])
    assert_equal [c.id], Company.includes(:contracts).where("contracts.id" => c.contracts.first).ids
  end
```

joshuay03 commented:
> I think we need to take some more inspiration from the existing pluck tests, as the new implementation is no longer covered by them.

@matthewd I can get a separate PR up with improved test coverage if that's alright?
